### PR TITLE
Add VcPkg support.

### LINKED ISSUE #172: Build system architecture for packages
Our current build system builds the entire SDK from the root of the repo. Given that we're going to publish individual packages it might make sense to be able to build packages independently. 

A couple ideas discussed 

* Build with cmake from the root and use flags to specify individual libraries to build
* Libraries can be built independently at either the service or library level

### LINKED ISSUE #526: Release Azure Core SDK with Vcpkg.

### LINKED ISSUE #575: Enable vcpkg publishing on azure-template project

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #526 
Closes #172 
Closes #575 
image
image
image

### CODE REVIEW

File: sdk/storage/azure-storage-files-shares/CMakeSettings.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      "variables": [
+        {
+          "name": "VCPKG_TARGET_TRIPLET",
+          "value": "x86-windows-static",
+          "type": "STRING"
+        }
+      ]
+    }
+  ]
+}
```
Comment (RickWinter) [inline sdk/storage/azure-storage-files-shares/CMakeSettings.json:44]:
```suggestion
}

```

File: sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in [hunk 2]
```diff
@@ -0,0 +1,8 @@
+@PACKAGE_INIT@
+
+include(CMakeFindDependencyMacro)
+find_dependency(azure-storage-common-cpp)
```
Comment (RickWinter) [inline sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in:4]: Why the -cpp at the end?

File: sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+cmake_minimum_required (VERSION 3.13)
+
+target_sources(
+    azure-storage-sample
+    PRIVATE
+    file_share_getting_started.cpp
+)
+
+target_link_libraries(azure-storage-sample PUBLIC Azure::azure-storage-files-shares)
```
Comment (RickWinter) [inline sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt:12]: This is odd, why isn't it Azure::Storage::Files::Shares?

File: sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in (hunk 2 above)
Comment (antkmsft) [reply sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in:4]: It is something for us to choose, we can get rid of it. But, to an extent, this is somewhat common. See RED - ports that are not in this PR, YELLOW are. I think, `-cpp` prefix is a good thing to prevent collision between libs from `azure-sdk-for-cpp` and others (for example, `-c`). But we can remove.
image

File: sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt (hunk 3 above)
Comment (antkmsft) [reply sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt:12]: This is not what you think it is (NOT a C++ namespace), and this is not how it works, in my understanding - you are only allowed to have one "namespace".

File: cmake-modules/az_vcpkg.cmake [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  elseif(DEFINED ENV{VCPKG_INSTALLATION_ROOT} AND NOT DEFINED CMAKE_TOOLCHAIN_FILE)
+    set(CMAKE_TOOLCHAIN_FILE "$ENV{VCPKG_INSTALLATION_ROOT}/scripts/buildsystems/vcpkg.cmake"
+        CACHE STRING "")
+  endif()
+  if(DEFINED ENV{VCPKG_DEFAULT_TRIPLET} AND NOT DEFINED VCPKG_TARGET_TRIPLET)
+    set(VCPKG_TARGET_TRIPLET "$ENV{VCPKG_DEFAULT_TRIPLET}" CACHE STRING "")
+  endif()
+endmacro()
+
+macro(az_vcpkg_export)
```
Comment (vhvb1989) [inline cmake-modules/az_vcpkg.cmake:20]: I would either use a parameter for the macro like `macro(az_vcpkg_export TARGET_NAME)` or a function with arguments instead of relying on the ${TARGET_NAME} to be set before calling this macro.

File: cmake-modules/az_vcpkg.cmake [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+        INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+    )
+
+  install(
+    EXPORT "${TARGET_NAME}-cppTargets"
+      DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/${TARGET_NAME}-cpp"
+      NAMESPACE Azure::
+      FILE "${TARGET_NAME}-cppTargets.cmake"
+    )
```
Comment (vhvb1989) [inline cmake-modules/az_vcpkg.cmake:43]: I would really love some comments for what each install is taking care of. If possible.
That would help on future maintenance.

File: sdk/core/azure-core/CMakeLists.txt [hunk 6]
```diff
@@ -53,16 +63,23 @@ add_library (
...
 # make sure that users can consume the project as a library.
-add_library (Azure::Core ALIAS ${TARGET_NAME})
+add_library(Azure::${TARGET_NAME} ALIAS ${TARGET_NAME})
 
 # coverage. Has no effect if BUILD_CODE_COVERAGE is OFF
 create_code_coverage(core ${TARGET_NAME} "${TARGET_NAME}-test")
 
 target_include_directories(${TARGET_NAME} PUBLIC ${CURL_INCLUDE_DIRS})
-target_link_libraries(${TARGET_NAME} PRIVATE CURL::libcurl Threads::Threads)
+target_link_libraries(${TARGET_NAME} PUBLIC CURL::libcurl Threads::Threads)
```
Comment (vhvb1989) [inline sdk/core/azure-core/CMakeLists.txt:80]: why are you changing this?
libcurl and threads libraries are consumed only by Azure CORE.  By doing PUBLIC, we are saying that the Azure SDK also provides this 2 with it

File: sdk/core/azure-core/CMakeLists.txt [hunk 7]
```diff
@@ -53,16 +63,23 @@ add_library (
...
 
 get_az_version("${CMAKE_CURRENT_SOURCE_DIR}/inc/azure/core/version.hpp")
-generate_documentation(azure-core ${AZ_LIBRARY_VERSION})
+generate_documentation(${TARGET_NAME} ${AZ_LIBRARY_VERSION})
```
Comment (vhvb1989) [inline sdk/core/azure-core/CMakeLists.txt:83]: At some point we got feedback from Charlie about not following the ${TARGET_NAME} approach and repeat the same words multiple times better

File: sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in (hunk 2 above)
Comment (antkmsft) [reply sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in:4]: FWIW, Google and AWS do also have `-cpp` suffix:
image

**Review by vhvb1989 — COMMENTED** [review]
Partial review.. WIll continue later

File: sdk/core/azure-core/CMakeLists.txt (hunk 6 above)
Comment (antkmsft) [reply sdk/core/azure-core/CMakeLists.txt:80]: But it does! We require customer to link with it.

File: sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in (hunk 2 above)
Comment (vhvb1989) [reply sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in:4]: It makes sense if Vcpkg can be used for something else than cpp..  Otherwise, it should be understood that it is all cpp.
libcurl and cmocka don't follow this (libcurl-cpp, cmocka-cpp)

Maybe these libraries use it because they have other languages support right?  (libcurl is only cpp, for instance)

I think the -cpp is fine

File: cmake-modules/az_vcpkg.cmake [hunk 8]
```diff
@@ -0,0 +1,69 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply cmake-modules/az_vcpkg.cmake:1]: We also have `az_version.cmake` there. I just used it as prefix. But we can change the prefix.

File: CMakeLists.txt [hunk 9]
```diff
@@ -33,7 +35,7 @@ endif()
 include(DefinePlatform)
 
 # Project definition
-project(az LANGUAGES CXX)
+project("azure-sdk" LANGUAGES CXX)
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply CMakeLists.txt:38]: We do, but we can remove it. This is not the goal of this PR.
I think having root file is still convenient, because otherwise you need other azure packages being installed and you consume them from the vcpkg, which could be an inconvenience if you need features that were not yet released, so then you would need to fork vcpkg and do what I did to test this, and this is too much time consuming.

File: sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt (hunk 3 above)
Comment (vhvb1989) [reply sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt:12]: It is an CMake target ALIAS,  it can also be Azure::Storage::Files::Shares, but usually people use only one groouping label + the CMake target name. The `Azure` is the grouping label, and then there is the name of the CMake target.
When this ALIAS was created looks like:
add_library(azure::storage::files::shares ALIAS azure-storage-files-shares), so it should work using it like that

File: sdk/core/azure-core/CMakeLists.txt (hunk 7 above)
Comment (antkmsft) [reply sdk/core/azure-core/CMakeLists.txt:83]: No strong opinion on this, but I find `${TARGET_NAME}` to be less error-prone.

File: sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt (hunk 3 above)
Comment (antkmsft) [reply sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt:12]: But this is not the target name that VcPkg is going to produce. VcPkg is going to tell you
```
The package azure-storage-files-shares-cpp:x64-windows-static provides CMake targets:

    find_package(azure-storage-files-shares-cpp CONFIG REQUIRED)
    target_link_libraries(main PRIVATE Azure::azure-storage-files-shares)
```

So then you have two different names you need to reference - one is alias `Azure::Storage::Files::Shares` which is good and valid when you build the entire SDK as a single project, and then you need an if() to provide `Azure::azure-storage-files-shares` as link target, when building as individual library (when VcPkg is getting built)

File: sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in (hunk 2 above)
Comment (antkmsft) [reply sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in:4]: One correction, libcurl is C, not Cpp :)

**Comment by antkmsft** [comment]
Earlier, with the Storage Team, we had a conversation about producing dynamic vs static libs.

I was able to build the package in all these official triplets:
* x64-windows (DLL, dynamic runtime)
* x64-windows-static (static lib, static runtime)
* x86-windows (DLL, dynamic runtime)

And these community triplets:
* x86-windows-static (static lib, static runtime)
* x86-windows-static-md (static lib, dynamic runtime)
* x64-windows-static-md (static lib, dynamic runtime)

I can see that DLLs are produced in the corresponding configurations (/bin directory inside the package directory + corresponding lib files in the /lib directory). We probably need to do a bit more verifications, but so far it looks like we are where we wanted to be - we don't dictate which type of lib to produce, we can be built into both.

So then, the question comes to updating the docs and CI/CD build matrix.

File: cmake-modules/az_vcpkg.cmake (hunk 4 above)
Comment (antkmsft) [reply cmake-modules/az_vcpkg.cmake:20]: We still need project name and version to be set as some variable, so that these commands work:
```
configure_file("${CMAKE_CURRENT_SOURCE_DIR}/vcpkg/CONTROL" "${CMAKE_BINARY_DIR}/vcpkg/ports/${TARGET_NAME}-cpp/CONTROL" @ONLY)
configure_file("${CMAKE_CURRENT_SOURCE_DIR}/vcpkg/portfile.cmake" "${CMAKE_BINARY_DIR}/vcpkg/ports/${TARGET_NAME}-cpp/portfile.cmake" @ONLY)
```

File: sdk/identity/azure-identity/vcpkg/CONTROL [hunk 10]
```diff
@@ -0,0 +1,7 @@
+Source: @TARGET_NAME@-cpp
+Version: @AZ_LIBRARY_VERSION@
+Build-Depends: azure-core-cpp
+Description: Microsoft Azure Identity SDK for C++
+  This library provides common authentication-related abstractions for Azure SDK.
+Homepage: https://github.com/Azure/azure-sdk-for-cpp
```
Comment (vhvb1989) [inline sdk/identity/azure-identity/vcpkg/CONTROL:6]: Point to Identity directly instead of the general SDK.

File: sdk/storage/CMakeLists.txt [hunk 11]
```diff
@@ -10,30 +10,26 @@ set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
...
-
-add_executable(azure-storage-sample)
-if(BUILD_STORAGE_SAMPLES)
-    set_target_properties(azure-storage-sample PROPERTIES EXCLUDE_FROM_ALL FALSE)
-else()
-    set_target_properties(azure-storage-sample PROPERTIES EXCLUDE_FROM_ALL TRUE)
+    add_subdirectory(azure-storage-common/test)
+    add_subdirectory(azure-storage-blobs/test)
+    add_subdirectory(azure-storage-files-datalake/test)
+    add_subdirectory(azure-storage-files-shares/test)
```
Comment (vhvb1989) [inline sdk/storage/CMakeLists.txt:25]: This looks like another candidate for creating an independent PR.
It looks like this is refactoring the Azure Storage tests and samples build structure, which makes me happy!, but should be committed as a different story

File: sdk/storage/azure-storage-blobs/CMakeLists.txt [hunk 12]
```diff
@@ -2,72 +2,57 @@
...
+
+target_include_directories(
+  ${TARGET_NAME}
+    PUBLIC
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/inc>
+      $<INSTALL_INTERFACE:include>
+      ${azure-storage-common-cpp_INCLUDE_DIRS}
+  )
+
+target_link_libraries(${TARGET_NAME} PUBLIC Azure::azure-storage-common)
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/CMakeLists.txt:53]: Is azure-storage-common expected to be used by end-users?
If not, this should be INTERFACE or PRIVATE

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/append_blob_client.hpp [hunk 13]
```diff
@@ -3,7 +3,7 @@
 
 #pragma once
 
-#include "azure/identity/client_secret_credential.hpp"
+#include "azure/core/credentials.hpp"
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/append_blob_client.hpp:6]: Is this related to vcpkg?

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp [hunk 14]
```diff
@@ -3,7 +3,7 @@
 
 #pragma once
 
-#include "azure/identity/client_secret_credential.hpp"
+#include "azure/core/credentials.hpp"
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:6]: I noticed this is applied to a lot of files. 
Please move this out of this PR to its own change

File: sdk/storage/azure-storage-blobs/test/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+
+cmake_minimum_required (VERSION 3.13)
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/test/CMakeLists.txt:4]: yes, I like this test and sample refactoring for Storage. But it needs to be a separated and visible change in our records.
Let's not smuggle it with the adding packages to vcpkg

File: sdk/storage/azure-storage-blobs/vcpkg/CONTROL [hunk 16]
```diff
@@ -0,0 +1,7 @@
+Source: @TARGET_NAME@-cpp
+Version: @AZ_LIBRARY_VERSION@
+Build-Depends: azure-storage-common-cpp
+Description: Microsoft Azure Storage Blobs SDK for C++
+  This library provides Azure Storage Blobs SDK.
+Homepage: https://github.com/Azure/azure-sdk-for-cpp
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/vcpkg/CONTROL:6]: Point to blobs path directly instead of the main SDK page

File: sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt (hunk 3 above)
Comment (vhvb1989) [reply sdk/storage/azure-storage-files-shares/sample/CMakeLists.txt:12]: You are getting
```
target_link_libraries(main PRIVATE Azure::azure-storage-files-shares)
```

Because when you created the ALIAS, you used `Azure::${TARGET_NAME}` which translates to Azure::azure-storage-files-shares.   The alias is expected be more like `Azure::Storage::Files::Shares`

File: sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in (hunk 2 above)
Comment (vhvb1989) [reply sdk/storage/azure-storage-files-shares/vcpkg/Config.cmake.in:4]: Yes, my bad! lol
Then that's the quick reason right?   Vcpkg contains both C and CPP packages. We want to add the -cpp to make it clear for users about it when using `./vcpkg search`

**Review by vhvb1989 — COMMENTED** [review]
There are changes unrelated to the vcpkg story.
My recommendation is to split this PR into at least 4 changes
-  Storage Test and Samples build refactoring.
-  Storage credentials update.
-  Core headers (context) update.
-  Core WIN32 ifdef update.

Let's not make those changes to become part of VCPKG story

File: sdk/storage/azure-storage-blobs/vcpkg/CONTROL (hunk 16 above)
Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/vcpkg/CONTROL:6]: Hmm. Does that risk leading to a 404 link if we restructure the repo, or will we always have the blobs path be stable?

File: sdk/storage/azure-storage-blobs/vcpkg/Config.cmake.in [hunk 17]
```diff
@@ -0,0 +1,8 @@
+@PACKAGE_INIT@
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/vcpkg/Config.cmake.in:1]: Do we need license headers for all these files?

File: sdk/storage/azure-storage-blobs/CMakeLists.txt (hunk 12 above)
Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/CMakeLists.txt:53]: > Is azure-storage-common expected to be used by end-users?

I think so. @Jinming-Hu, @katmsft - can you confirm what the intent is for azure-storage-common around end-user usage?

There are certain things like StorageException/etc. that are in azure-storage-common, so I suspect folks will reference it (probably not directly though). This package **doesn't only** contain common utility types/methods used by the other three storage packages.

File: cmake-modules/az_vcpkg.cmake (hunk 4 above)
Comment (antkmsft) [reply cmake-modules/az_vcpkg.cmake:21]: This is not only about the ${TARGET_NAME}, but see sdk/core/azure-core/vcpkg/Config.cmake.in:
```
find_dependency(nlohmann_json @NLOHMANN_JSON_MIN_REQUIRED_VERSION@)
if(@BUILD_TRANSPORT_CURL@)
  find_dependency(CURL @CURL_MIN_REQUIRED_VERSION@)
endif()
```

We just can't pass all the possible variables as parameters.

File: sdk/core/azure-core/CMakeSettings.json [hunk 18]
```diff
@@ -0,0 +1,44 @@
+﻿{
+  "configurations": [
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/core/azure-core/CMakeSettings.json:2]: But we have the same file checked in for the root directory. If we keeping that, let's have this one too. Because it is convenient to load just a single directory, build, and verify that it builds in isolation, which is important for vcpkg.
Not the end of the world, this file can be deleted and recreated, but I think then the same file in the root directory needs to go as well. But I think having this file is rather a convenience. Let me know, I have no strong opinion on this.

File: .gitignore [hunk 19]
```diff
@@ -339,3 +339,5 @@ build/
 # vscode
 .vscode/
 .factorypath
+
+out/
```
Comment (vhvb1989) [inline .gitignore:343]: this is already covered by the first lines. We cover all Visual Studio add-ons

File: cmake-modules/AddGoogleTest.cmake [hunk 20]
```diff
@@ -67,7 +67,7 @@ endif()
 
 # Target must already exist
 macro(add_gtest TESTNAME)
-    target_link_libraries(${TESTNAME} PUBLIC gtest gmock gtest_main)
+    target_link_libraries(${TESTNAME} PRIVATE gtest gmock gtest_main)
```
Comment (vhvb1989) [inline cmake-modules/AddGoogleTest.cmake:70]: this is good!  thank you!

File: cmake-modules/az_vcpkg.cmake (hunk 4 above)
Comment (vhvb1989) [reply cmake-modules/az_vcpkg.cmake:21]: We need to pass the required variables as parameters.
The targetName is required because if it is not set before calling this macro, the macro will still run using an empty string as targetName. Or even worst, it could take the last value set to targetName which can be not the one we need, but a previously used.

For any variable that starts with `CMake...` we don't need to pass it as parameter because we know CMake will take care of assigning those values.

My approach is, if you are using a global constant variable, then it's fine (like the CMake vairables or maybe the AZ_ROOT_DIR) because they are set once and won't change.
But if you are using a variable that is expected to be different everytime this macro is called, then we need it as parameter and not as a global variable

File: sdk/core/azure-core/CMakeLists.txt [hunk 21]
```diff
@@ -64,17 +73,25 @@ add_library (
...
+target_include_directories(
+  ${TARGET_NAME}
+    PUBLIC
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/inc>
+      $<INSTALL_INTERFACE:include>
+  )
 
 # make sure that users can consume the project as a library.
-add_library (Azure::Core ALIAS ${TARGET_NAME})
+add_library(Azure::${TARGET_NAME} ALIAS ${TARGET_NAME})
```
Comment (vhvb1989) [inline sdk/core/azure-core/CMakeLists.txt:84]: I don't think this is OK.  the Alias should be Azure::Core and not Azure::azure-core

File: sdk/identity/azure-identity/CMakeSettings.json [hunk 22]
```diff
@@ -0,0 +1,44 @@
+﻿{
+  "configurations": [
```
Comment (vhvb1989) [inline sdk/identity/azure-identity/CMakeSettings.json:2]: Committing this file breaks other developers.
As a Developer, I can set up my VIsual Studio settings with some configurations I want to use where I set this and that option.
Then, after getting last master, It would break my configuration.

I insist that pushing this file ends up to be more problematic.
If the intention is for first-time-users to open Visual Studio and get some initial configuration, that will work, but later, once they learn more and start changing that configuration, will they commit changes to this file?  

Also, This changes are not required at all for vcpkg support.
If you feel strong about sharing this configuration, please move it to another PR

**Review by vhvb1989 — COMMENTED** [review]
Still having some comments to be reviewed around there.

Thank you very much for moving the other changes out of this PR. This is looking better now!

Please consider moving the CMakeSettings.json files to another PR.

Please consider moving the CMake files changes that changes the build structure from a main root file to enable individual CMake files for each package to be built to another PR

Then this PR would only show the vcpkg addition perfectly!

File: sdk/core/azure-core/CMakeLists.txt [hunk 23]
```diff
@@ -100,18 +103,21 @@ target_include_directories(
   azure-core
     PUBLIC
       $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/inc>
-      $<INSTALL_INTERFACE:include/az_core>
+      $<INSTALL_INTERFACE:include>
 )
 
 # make sure that users can consume the project as a library.
-add_library(Azure::Core ALIAS azure-core)
+add_library(Azure::azure-core ALIAS azure-core)
```
Comment (vhvb1989) [inline sdk/core/azure-core/CMakeLists.txt:110]: what happen if we leave it as Azure::Core ?  
Looks weird to repeat azure twice

**Review by vhvb1989 — APPROVED** [review]
Looks great! thank you.

Please elaborate about why do we need to repeat label `azure` for the ALIAS like
   `add_library(Azure::azure-template ALIAS azure-template)`

Also, please wait at least for a second approval from the team before merging.

It looks good and ready to me!  but I wouldn't like to be the only reviewer for this change.
Thank you!

File: sdk/core/azure-core/CMakeLists.txt [hunk 24]
```diff
@@ -100,23 +103,26 @@ target_include_directories(
...
 
 # coverage. Has no effect if BUILD_CODE_COVERAGE is OFF
-create_code_coverage(core azure-core "azure-core-test")
+create_code_coverage(core azure-core azure-core-test)
 
+# Needs Azure::azure-storage-blobs be public as long as we #include<curl.h> in public headers.
```
Comment (antkmsft) [inline sdk/core/azure-core/CMakeLists.txt:115]:
```suggestion
# ${CURL_INCLUDE_DIRS} needs to be public as long as we #include<curl.h> in public headers.
```

File: sdk/storage/azure-storage-common/CMakeLists.txt [hunk 25]
```diff
@@ -56,22 +66,34 @@ set(
...
+
+target_include_directories(azure-storage-common INTERFACE ${LIBXML2_INCLUDE_DIRS})
+target_link_libraries(azure-storage-common PUBLIC Azure::azure-core)
+target_link_libraries(azure-storage-common INTERFACE Threads::Threads ${LIBXML2_LIBRARIES})
 
-target_include_directories(azure-storage-common PUBLIC inc ${LIBXML2_INCLUDE_DIRS})
-target_link_libraries(azure-storage-common Threads::Threads azure-core ${LIBXML2_LIBRARIES})
 if(MSVC)
-    target_link_libraries(azure-storage-common bcrypt)
+    target_link_libraries(azure-storage-common INTERFACE bcrypt)
```
Comment (RickWinter) [inline sdk/storage/azure-storage-common/CMakeLists.txt:84]: bcrypt is a dependency, why is it not listed in the package dependencies?

File: sdk/storage/azure-storage-blobs/vcpkg/CONTROL [hunk 26]
```diff
@@ -0,0 +1,7 @@
+Source: azure-storage-blobs-cpp
+Version: @AZ_LIBRARY_VERSION@
+Build-Depends: azure-storage-common-cpp
```
Comment (RickWinter) [inline sdk/storage/azure-storage-blobs/vcpkg/CONTROL:3]: This package also depends on azure-core.

Do we need to list bcrypt, LibXml2, ...  and all other dependencies?

File: sdk/storage/azure-storage-common/vcpkg/CONTROL [hunk 27]
```diff
@@ -0,0 +1,7 @@
+Source: azure-storage-common-cpp
+Version: @AZ_LIBRARY_VERSION@
+Build-Depends: azure-core-cpp, libxml2, openssl (!windows)
```
Comment (RickWinter) [inline sdk/storage/azure-storage-common/vcpkg/CONTROL:3]: bcrypt, LibXML2

File: sdk/core/azure-core/CMakeLists.txt (hunk 23 above)
Comment (antkmsft) [reply sdk/core/azure-core/CMakeLists.txt:110]: It tells me that Azure::Core is an alias and can't be exported. If I simply use it in this specific place, the target library vcpkg produces is still Aaure::azure-core. So then when you build another lib as vcpackage, and we find_package() and then try to reference it as Azure::Core, it fails to find it.

File: sdk/storage/azure-storage-blobs/vcpkg/CONTROL (hunk 26 above)
Comment (antkmsft) [reply sdk/storage/azure-storage-blobs/vcpkg/CONTROL:3]: storage-common depends on azure-core and brings these dependencies, it is not necessary to list them here, I can confirm it all builds when you have nothing installed and request azure-storage-common right away,

File: sdk/storage/azure-storage-common/CMakeLists.txt (hunk 25 above)
Comment (antkmsft) [reply sdk/storage/azure-storage-common/CMakeLists.txt:84]: bcrypt is a Win API header and not a package: https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/
From the perspective of vcpkg, it is not a dependency we are bringing.

File: sdk/storage/azure-storage-common/vcpkg/CONTROL (hunk 27 above)
Comment (antkmsft) [reply sdk/storage/azure-storage-common/vcpkg/CONTROL:3]: libxml2 is listed, bcrypt is WinAPI and not vcpkg.

File: cmake-modules/az_vcpkg.cmake [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+  configure_file("${CMAKE_CURRENT_SOURCE_DIR}/vcpkg/portfile.cmake" "${CMAKE_BINARY_DIR}/vcpkg/ports/${targetName}-cpp/portfile.cmake" @ONLY)
+
+  # Standard names for folders such as "bin", "lib", "include". We could hardcode, but some other libs use it too (curl).
+  include(GNUInstallDirs)
+
+  # When installing, copy our "inc" directory (headers) to "include" directory at the install location.
+  install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/inc/" DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+
+  # Copy license as "copyright" (vcpkg dictates naming and location).
+  install(FILES "${AZ_ROOT_DIR}/LICENSE.txt" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/${targetName}-cpp" RENAME "copyright")
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply cmake-modules/az_vcpkg.cmake:36]: I haven't seen other packages bringing them. Also, this change brings what is essential to vcpkg to build, without the license file validation does fail.

File: sdk/core/azure-core/CMakeLists.txt [hunk 29]
```diff
@@ -10,12 +10,15 @@ set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
...
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../../../cmake-modules")
 
+include(az_vcpkg)
 include(az_version)
 include(CodeCoverage)
 include(DefineTransportAdapter)
 include(doxygen_common)
 include(global_compile_options)
 
+az_vcpkg_integrate()
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/core/azure-core/CMakeLists.txt:20]: I agree, but this PR would be the first that uses Pascal - version does not, for example, and all others I can see. Let's unify all them to a consistent style in another PR.
